google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiManager.java

```diff
@@ -119,6 +146,175 @@ static void enableApis(Set<CloudLibrary> libraries, CloudProject cloudProject, P
+      ServiceAccount serviceAccount,
+      Set<Role> roles,
+      CloudProject cloudProject)
+      throws IOException {
+    CloudResourceManager resourceManager =
+        GoogleApiClientFactory.getInstance().getCloudResourceManagerClient(user.getCredential());
+
+    Policy existingPolicy =
+        resourceManager
+            .projects()
+            .getIamPolicy(cloudProject.projectId(), new GetIamPolicyRequest())
+            .execute();
+    List<Binding> bindings = Lists.newArrayList(existingPolicy.getBindings());
+
+    List<Binding> additionalBindings =
+        roles
+            .stream()
+            .map(
+                role -> {
+                  Binding binding = new Binding();
+                  binding.setRole(role.getName());
+                  binding.setMembers(
+                      ImmutableList.of(
+                          SERVICE_ACCOUNT_ROLE_REQUEST_PREFIX + serviceAccount.getEmail()));
+                  return binding;
+                })
+            .collect(Collectors.toList());
+
+    bindings.addAll(additionalBindings);
+
+    SetIamPolicyRequest policyRequest = new SetIamPolicyRequest();
+    Policy newPolicy = new Policy();
+    newPolicy.setBindings(bindings);
+    policyRequest.setPolicy(newPolicy);
+
+    resourceManager.projects().setIamPolicy(cloudProject.projectId(), policyRequest).execute();
+  }
+
+  /**
+   * Using the supplied {@link ServiceAccount}, this creates and returns a new {@link
+   * ServiceAccountKey}.
+   */
+  private static ServiceAccountKey createServiceAccountKey(
+      CredentialedUser user, ServiceAccount serviceAccount) throws IOException {
+    Iam iam = GoogleApiClientFactory.getInstance().getIamClient(user.getCredential());
+
+    CreateServiceAccountKeyRequest keyRequest = new CreateServiceAccountKeyRequest();
+    return iam.projects()
+        .serviceAccounts()
+        .keys()
+        .create(serviceAccount.getName(), keyRequest)
+        .execute();
+  }
+
+  /**
+   * Writes the service account private key data in JSON form to the filesystem. The private key
+   * itself is contained within the {@link ServiceAccountKey} returned from the IAM API. The json
+   * key is encoded within {@link ServiceAccountKey#getPrivateKeyData()} in base64.
+   *
+   * @param key the {@link ServiceAccountKey} containing the base64 encoding of the json private key
+   * @param downloadDir the {@link Path} on the file system to download the ky
+   * @param cloudProject the {@link CloudProject} associated with this service account
+   * @return the {@link Path} to the service account key that was written
+   * @throws IOException if the an IO error occurs when writing the file
+   */
+  private static Path writeServiceAccountKey(
+      ServiceAccountKey key, Path downloadDir, CloudProject cloudProject) throws IOException {
+    Path keyPath =
+        Paths.get(
+            downloadDir.toString(), cloudProject.projectName() + "-" + getTimestamp() + ".json");
```

Comment:
What if we consider using actual service account name for a file name (i.e. name we generated from `createServiceAccountId()`?

Reply:
yeah that sounds reasonable. I went with the cloud project name since that is probably the most immediately recognizable name to the user. But it creates ambiguity when there are multiple service accounts. 

What about a combination of the cloud project name and the service account name?

Reply:
Combination sounds good to me, could be a bit long, but least amount of ambiguity for sure.

Reply:
I decided to keep cloud project as the name of the key for the following reasons:
- its too long with cloud project and service account, like you said
- the web console also generates keys with the same strategy (cloud project name + some unique id).

sound ok to you?

Reply:
Sounds good.